File: Reqnroll/Chapter16/16.1-ScenarioResourcePattern/Before/WIMP.Specs/Drivers/PaymentApiDriver.cs
using WIMP.App.Models;
using WIMP.Specs.Support;

namespace WIMP.Specs.Drivers;

public class PaymentApiDriver(RestApiContext restApiContext)
{
    public TestAction<Payment> AuthorizePaymentFor(int orderNo) =>
        new LambdaAction<Payment>("Authorize payment", async () =>
            await restApiContext.ProcessRequest<Payment>(
                "Authorize payment", HttpMethod.Post, $"/api/payment/authorize?orderNo={orderNo}"));
}

File: CucumberJVM/Chapter20/20.2-TestConfigurationProviderPattern/After/src/test/java/com/wimp/app/specs/drivers/PaymentApiDriver.java
package com.wimp.app.specs.drivers;

import com.wimp.app.models.Payment;
import com.wimp.app.specs.support.LambdaAction;
import com.wimp.app.specs.support.TestAction;
import org.springframework.stereotype.Component;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.service.annotation.HttpExchange;
import org.springframework.web.service.annotation.PostExchange;

@Component
public class PaymentApiDriver {
    private final PaymentApiClient paymentApiClient;

    @HttpExchange("/api/payment")
    public interface PaymentApiClient {
        @PostExchange("/authorize")
        Payment authorize(@RequestParam int orderNo);
    }

    @SuppressWarnings("SpringJavaInjectionPointsAutowiringInspection")
    public PaymentApiDriver(PaymentApiClient paymentApiClient) {
        this.paymentApiClient = paymentApiClient;
    }

    public TestAction<Payment> authorizePaymentFor(int orderNo) {
        return new LambdaAction<>("Authorize payment", orderNo, () -> paymentApiClient.authorize(orderNo));
    }
}
